simplemma/language_detector.py

```diff
@@ -1,128 +1,179 @@
-    ) -> None:
-        self.tokenizer = tokenizer
-        self.max_tokens = max_tokens
-        self.capitalized_threshold = capitalized_threshold
-
-    def sample_tokens(self, text: str) -> List[str]:
-        """Extract potential tokens, scramble them, potentially get rid of capitalized
-        ones, and return the most frequent."""
-
-        counter = Counter(token for token in self.tokenizer.split_text(text))
-
-        if self.capitalized_threshold > 0:
-            deletions = [token for token in counter if token[0].isupper()]
-            if len(deletions) < self.capitalized_threshold * len(counter):
-                for token in deletions:
-                    del counter[token]
-
-        return [item[0] for item in counter.most_common(self.max_tokens)]
-
-
-class RelaxedTokenSampler(TokenSampler):
-    def __init__(self) -> None:
-        super().__init__(
-            tokenizer=Tokenizer(RELAXED_SPLIT_INPUT),
-            max_tokens=1000,
-            capitalized_threshold=0,
-        )
-
-
 def in_target_language(
     text: str,
     lang: Optional[Union[str, Tuple[str, ...]]] = None,
+    greedy: bool = False,
     dictionary_factory: DictionaryFactory = DictionaryFactory(),
     token_sampler: TokenSampler = TokenSampler(),
 ) -> float:
     """Determine which proportion of the text is in the target language(s)."""
-    total = 0
-    in_target = 0
-    dictionaries = dictionary_factory.get_dictionaries(lang)
-    for token in token_sampler.sample_tokens(text):
-        total += 1
-        for lang_code, lang_dictionary in dictionaries.items():
-            candidate = _return_lemma(
-                token, lang_dictionary, greedy=True, lang=lang_code
-            )
-            if candidate is not None:
-                in_target += 1
-                break
-    if total > 0 and in_target > 0:
-        return in_target / total
-    return 0
+    return LanguageDetector(dictionary_factory).detect_text_proportion_in_languages(
+        text, lang, greedy, token_sampler
+    )
 
 
-def _return_default() -> List[Tuple[str, float]]:
-    # todo: None if 'unk'?
-    return [("unk", 1)]
-
-
-def lang_detector(
+def _convert_results_to_sorted_list(
+    results: Dict[str, float]
+) -> List[Tuple[str, float]]:
+    list_results: List[Tuple[str, float]] = sorted(
+        results.items(), key=itemgetter(1), reverse=True
+    )
+    # switch unknown to the end
+    for i, item in enumerate(list_results):
+        if item[0] == "unk":
+            pair = list_results.pop(i)
+            list_results.append(pair)
+    return list_results
+
+
+def langdetect(
     text: str,
     lang: Optional[Union[str, Tuple[str, ...]]] = None,
     greedy: bool = False,
     dictionary_factory: DictionaryFactory = DictionaryFactory(),
-    token_sampler: TokenSampler = TokenSampler(),
-    backup_sampler: TokenSampler = RelaxedTokenSampler(),
+    token_samplers: List[TokenSampler] = [
+        TokenSampler(),
+        TokenSampler(
```

Comment:
This could still be defined above.

Reply:
Could you please address this, i.e. define a relaxed/backup sampler outside of the function?